drivers/flash/flash_stm32_ospi.c

```diff
@@ -0,0 +1,1505 @@
+		LOG_DBG("erase size will be used as layout page size");
+		layout_page_size = erase_size;
+	}
+
+	/* Warn but accept layout page sizes that leave inaccessible
+	 * space.
+	 */
+	if ((flash_size % layout_page_size) != 0) {
+		LOG_INF("layout page %u wastes space with device size %zu",
+			layout_page_size, flash_size);
+	}
+
+	data->layout.pages_size = layout_page_size;
+	data->layout.pages_count = flash_size / layout_page_size;
+	LOG_DBG("layout %u x %u By pages", data->layout.pages_count,
+					   data->layout.pages_size);
+
+	return rv;
+}
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+static int spi_nor_process_bfp(const struct device *dev,
+			       const struct jesd216_param_header *php,
+			       const struct jesd216_bfp *bfp)
+{
+	const struct flash_stm32_ospi_config *dev_cfg = DEV_CFG(dev);
+	struct flash_stm32_ospi_data *data = DEV_DATA(dev);
+	struct jesd216_erase_type *etp = data->erase_types;
+	const size_t flash_size = jesd216_bfp_density(bfp) / 8U;
+
+	if (flash_size != dev_cfg->flash_size) {
+		LOG_DBG("Unexpected flash size: %u", flash_size);
+	}
+
+	LOG_DBG("%s: %u MiBy flash", dev->name, (uint32_t)(flash_size >> 20));
+
+	/* Copy over the erase types, preserving their order.  (The
+	 * Sector Map Parameter table references them by index.)
+	 */
+	memset(data->erase_types, 0, sizeof(data->erase_types));
+	for (uint8_t ti = 1; ti <= ARRAY_SIZE(data->erase_types); ++ti) {
+		if (jesd216_bfp_erase(bfp, ti, etp) == 0) {
+			LOG_DBG("Erase %u with %02x",
+					(uint32_t)BIT(etp->exp), etp->cmd);
+		}
+		++etp;
+	}
+
+	data->page_size = jesd216_bfp_page_size(php, bfp);
+
+	LOG_DBG("Page size %u bytes", data->page_size);
+	LOG_DBG("Flash size %u bytes", flash_size);
+	return 0;
+}
+
+static int flash_stm32_ospi_init(const struct device *dev)
+{
+	const struct flash_stm32_ospi_config *dev_cfg = DEV_CFG(dev);
+	struct flash_stm32_ospi_data *dev_data = DEV_DATA(dev);
+	uint32_t ahb_clock_freq;
+	uint32_t prescaler = 0;
+	int ret;
+
+	/* The SPI/DTR is not a valid config of data_mode/data_rate according to the DTS */
+	if ((dev_cfg->data_mode == OSPI_SPI_MODE)
+		&& (dev_cfg->data_rate == OSPI_DTR_TRANSFER)) {
+		/* already the right config, continue */
+		LOG_ERR("OSPI mode SPI/DTR is not valid");
+		return -ENOTSUP;
+	}
+
+	/* Signals configuration */
+	ret = pinctrl_apply_state(dev_cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		LOG_ERR("OSPI pinctrl setup failed (%d)", ret);
+		return ret;
+	}
+
+	/* Initializes the independent peripherals clock */
+	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_SYSCLK); /* */
+
+	/* Clock configuration */
+	if (clock_control_on(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
+			     (clock_control_subsys_t) &dev_cfg->pclken) != 0) {
+		LOG_DBG("Could not enable OSPI clock");
+		return -EIO;
+	}
+
+	if (clock_control_get_rate(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
+			(clock_control_subsys_t) &dev_cfg->pclken,
+			&ahb_clock_freq) < 0) {
+		LOG_DBG("Failed to get AHB clock frequency");
+		return -EIO;
```

Comment:
Note that this is correct only if SYSCLK rate equals AHB bus clock rate.

Reply:
> Note that this is correct only if SYSCLK rate equals AHB bus clock rate.

Could that be written as a comment?